src/drivers/roboclaw/RoboClaw.cpp

```diff
@@ -112,84 +135,147 @@ RoboClaw::~RoboClaw()
-	uint32_t count = 0;
-	uint8_t *countBytes = (uint8_t *)(&count);
-	countBytes[3] = rbuf[0];
-	countBytes[2] = rbuf[1];
-	countBytes[1] = rbuf[2];
-	countBytes[0] = rbuf[3];
-	uint8_t status = rbuf[4];
-	uint8_t checksum = rbuf[5];
-	uint8_t checksum_computed = (sum + _sumBytes(rbuf, 5)) &
-				    checksum_mask;
-
-	// check if checksum is valid
-	if (checksum != checksum_computed) {
-		printf("checksum failed: expected %d got %d\n",
-		       checksum, checksum_computed);
-		return -1;
-	}
+	memset((void *) &_wheelEncoderMsg, 0, sizeof(wheel_encoders_s));
+	_wheelEncoderMsg.timestamp = hrt_absolute_time();
+	_wheelEncodersAdv = orb_advertise(ORB_ID(wheel_encoders), &_wheelEncoderMsg);
+
+	while (!taskShouldExit) {
 
-	int overFlow = 0;
+		int pret = poll(fds, sizeof(fds) / sizeof(pollfd), waitTime / 1000);
 
-	if (status & STATUS_REVERSE) {
-		//printf("roboclaw: reverse\n");
+		// No timeout, update on either the actuator controls or the armed state
+		if (pret > 0 && (fds[0].revents & POLLIN || fds[1].revents & POLLIN)) {
+			orb_copy(ORB_ID(actuator_controls_0), _actuatorsSub, &_actuatorControls);
+			orb_copy(ORB_ID(actuator_armed), _armedSub, &_actuatorArmed);
+
+			int drive_ret = 0, turn_ret = 0;
+
+			// Disarmed
+			if (!_actuatorArmed.armed || _actuatorArmed.lockdown || _actuatorArmed.manual_lockdown
+			    || _actuatorArmed.force_failsafe) {
+				// If disarmed, I want to be certain that the stop command gets through.
+				while ((drive_ret = drive(0.0)) <= 0 || (turn_ret = turn(0.0)) <= 0) {
+					PX4_ERR("Error trying to stop: Drive: %d, Turn: %d", drive_ret, turn_ret);
+				}
+
+			} else {
+				drive_ret = drive(_actuatorControls.control[actuator_controls_s::INDEX_THROTTLE]);
+				turn_ret = turn(_actuatorControls.control[actuator_controls_s::INDEX_YAW]);
+
+				if (drive_ret <= 0 || turn_ret <= 0) {
+					PX4_ERR("Error controlling RoboClaw. Drive err: %d. Turn err: %d", drive_ret, turn_ret);
+				}
+			}
+
+			// A timeout occurred, which means that it's time to update the encoders
+
+		} else {
+			encoderTaskLastRun = hrt_absolute_time();
+
+			if (readEncoder() > 0) {
+				_wheelEncoderMsg.timestamp = encoderTaskLastRun;
+				_wheelEncoderMsg.encoder_position[0] = _encoderCounts[0];
+				_wheelEncoderMsg.encoder_position[1] = _encoderCounts[1];
+
+				//PX4_INFO("[%llu] PUBLISHING", _wheelEncoderMsg.timestamp);
+				orb_publish(ORB_ID(wheel_encoders), _wheelEncodersAdv, &_wheelEncoderMsg);
+
+				//PX4_INFO("[%llu] Reading encoders", hrt_absolute_time());
+
+			} else {
+				PX4_ERR("Error reading encoders");
+			}
+		}
+
+		waitTime = ENCODER_READ_PERIOD_MS * 1000 - (hrt_absolute_time() - encoderTaskLastRun);
+		waitTime = waitTime < 0 ? 0 : waitTime;
 	}
 
-	if (status & STATUS_UNDERFLOW) {
-		//printf("roboclaw: underflow\n");
-		overFlow = -1;
+	orb_unsubscribe(_actuatorsSub);
+	orb_unsubscribe(_armedSub);
+	orb_unadvertise(_wheelEncodersAdv);
+}
+
+int RoboClaw::readEncoder()
+{
+
+	uint8_t rbuff[10];
```

Comment:
I'd put this magic 10 number into a `#define RBUFSIZE 10` or `static constexpr unsigned RBUFSIZE = 10`.